File: day2/day2.py
#!/usr/bin/env python3

def part1(program):
    x, depth = 0, 0

    for cmd, arg in program:
        if cmd == "forward":
            x += arg
        elif cmd == "down":
            depth += arg
        elif cmd == "up":
            depth -= arg

    return x * depth

def part2(program):
    x, depth, aim = 0, 0, 0

    for cmd, arg in program:
        if cmd == "forward":
            x += arg
            depth += arg * aim
        elif cmd == "down":
            aim += arg
        elif cmd == "up":
            aim -= arg

    return x * depth

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            program = [(x[0], int(x[1])) for x in (x.split() for x in f.read().splitlines())]

        print("Part1:", part1(program))
        print("Part2:", part2(program))
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

File: day2/day2.cpp
#include <fstream>
#include <fmt/format.h>
#include <vector>

using namespace std;

enum class Cmd
{
	FORWARD,
	DOWN,
	UP
};

struct Ins
{
	Cmd cmd;
	int arg;
};

istream& operator>>(istream& in, Ins &instr)
{
	string s;
	if (!(in >> s))
	{
		return in;
	}
	else if (s == "down")
	{
		instr.cmd = Cmd::DOWN;
	}
	else if (s == "up")
	{
		instr.cmd = Cmd::UP;
	}
	else if (s == "forward")
	{
		instr.cmd = Cmd::FORWARD;
	}
	else
	{
		in.setstate(ios::failbit);
		return in;
	}
	return in >> instr.arg;
}

static int part1(const vector<Ins>& program)
{
	int x = 0;
	int depth = 0;

	for (auto [cmd, arg] : program)
	{
		switch (cmd)
		{
		case Cmd::FORWARD: x += arg; break;
		case Cmd::DOWN: depth += arg; break;
		case Cmd::UP: depth -= arg; break;
		}
	}

	return x * depth;
}

static int part2(const vector<Ins>& program)
{
	int x = 0;
	int depth = 0;
	int aim = 0;

	for (auto [cmd, arg] : program)
	{
		switch (cmd)
		{
		case Cmd::FORWARD:
			x += arg;
			depth += arg * aim;
			break;

		case Cmd::DOWN:
			aim += arg;
			break;

		case Cmd::UP:
			aim -= arg;
			break;
		}
	}

	return x * depth;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filaname>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<Ins> program;
	Ins tmp;
	while (in >> tmp)
	{
		program.emplace_back(tmp);
	}
	in.close();

	fmt::print("Part1: {}\n", part1(program));
	fmt::print("Part2: {}\n", part2(program));

	return 0;
}
